File: generate_plan.py
import csv
import random

questions = []
with open('d:/DSA/questions.csv', 'r', encoding='utf-8') as f:
    reader = csv.DictReader(f)
    for row in reader:
        questions.append(row)

easies = [q for q in questions if q['Difficulty Level'].strip() == 'Easy']
mediums = [q for q in questions if q['Difficulty Level'].strip() == 'Medium']
hards = [q for q in questions if q['Difficulty Level'].strip() == 'Hard']

# Shuffle each category for randomness
random.seed(42)  # For good measure, though not strictly required
random.shuffle(easies)
random.shuffle(mediums)
random.shuffle(hards)

with open('d:/DSA/study_plan.csv', 'w', newline='', encoding='utf-8') as f:
    writer = csv.writer(f)
    writer.writerow(['Day', 'Pattern', 'Difficulty Level', 'Problem Name'])
    
    day = 1
    while easies or mediums or hards:
        daily_q = []
        
        # We aim for a balanced daily diet: 2 Easy, 6 Medium, 2 Hard
        for _ in range(2):
            if easies: daily_q.append(easies.pop())
        for _ in range(6):
            if mediums: daily_q.append(mediums.pop())
        for _ in range(2):
            if hards: daily_q.append(hards.pop())
            
        # If we didn't reach 10 (because we ran out of a specific difficulty), 
        # fill the rest of the 10 spots with whatever is remaining.
        while len(daily_q) < 10 and (easies or mediums or hards):
            if mediums: daily_q.append(mediums.pop())
            elif easies: daily_q.append(easies.pop())
            elif hards: daily_q.append(hards.pop())
            
        # Shuffle within the day so it's not always in the exact same difficulty order
        random.shuffle(daily_q)
        
        for q in daily_q:
            writer.writerow([f'Day {day}', q['Pattern'], q['Difficulty Level'], q['Problem Name']])
        
        day += 1

print("Successfully generated d:/DSA/study_plan.csv")

File: generate_plan.js
const fs = require('fs');

const content = fs.readFileSync('d:/DSA/questions.csv', 'utf-8');
const lines = content.split(/\r?\n/).filter(line => line.trim().length > 0);

// skip header
const questions = lines.slice(1).map(line => {
    // split by comma, keeping the rest of the string if it contains commas
    const parts = line.split(',');
    const pattern = parts[0];
    const diff = parts[1];
    const name = parts.slice(2).join(',');
    return { pattern, diff, name };
});

const easies = questions.filter(q => q.diff === 'Easy');
const mediums = questions.filter(q => q.diff === 'Medium');
const hards = questions.filter(q => q.diff === 'Hard');

// Fisher-Yates shuffle
function shuffle(array) {
    for (let i = array.length - 1; i > 0; i--) {
        const j = Math.floor(Math.random() * (i + 1));
        [array[i], array[j]] = [array[j], array[i]];
    }
}

shuffle(easies);
shuffle(mediums);
shuffle(hards);

let csvOutput = 'Day,Pattern,Difficulty Level,Problem Name\n';
let day = 1;

while (easies.length > 0 || mediums.length > 0 || hards.length > 0) {
    let daily_q = [];
    
    for (let i = 0; i < 2; i++) {
        if (easies.length) daily_q.push(easies.pop());
    }
    for (let i = 0; i < 6; i++) {
        if (mediums.length) daily_q.push(mediums.pop());
    }
    for (let i = 0; i < 2; i++) {
        if (hards.length) daily_q.push(hards.pop());
    }
    
    while (daily_q.length < 10 && (easies.length || mediums.length || hards.length)) {
        if (mediums.length) daily_q.push(mediums.pop());
        else if (easies.length) daily_q.push(easies.pop());
        else if (hards.length) daily_q.push(hards.pop());
    }
    
    shuffle(daily_q);
    
    for (const q of daily_q) {
        csvOutput += `Day ${day},${q.pattern},${q.diff},${q.name}\n`;
    }
    day++;
}

fs.writeFileSync('d:/DSA/study_plan.csv', csvOutput);
console.log('Successfully generated d:/DSA/study_plan.csv');
